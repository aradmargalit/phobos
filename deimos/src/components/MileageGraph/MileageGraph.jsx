import { meanBy as _meanBy, startCase as _startCase } from 'lodash';
import moment from 'moment';
import React from 'react';

import IntervalGraph from '../IntervalGraph';

const average = data => _meanBy(data, 'miles');
const adjustedWeekNumber = ((moment().day() + 6) % 7) + 1;

const projection = (data, intervalType) => {
  const running = data[data.length - 1].miles;
  switch (intervalType) {
    case 'month':
      return moment().daysInMonth() * (running / moment(new Date()).date());
    case 'year':
      return 365 * (running / moment().dayOfYear());
    case 'week':
      return 7 * (running / adjustedWeekNumber);
    default:
      return 0;
  }
};

export default function MileageGraph({ loading, intervalData, intervalType }) {
  const data = intervalData.map(({ interval, miles }) => ({
    interval,
    miles: parseFloat(miles.toFixed(2)),
  }));

  const startCaseIntervalType = _startCase(intervalType);

  return (
    <IntervalGraph
      loading={loading}
      data={data}
      average={average(data)}
      projection={{ x: data[data.length - 1].interval, y: projection(data, intervalType) }}
      title={`${startCaseIntervalType}ly Workout Miles`}
      color="#d4504f"
      stroke="#912827"
      xAxisKey="interval"
      dataKey="miles"
      unit={startCaseIntervalType}
      tooltipFormatter={value => [`${value} Miles`, '']}
    />
  );
}
