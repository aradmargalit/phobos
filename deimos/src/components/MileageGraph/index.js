import MileageGraph from './MileageGraph';

export default MileageGraph;
