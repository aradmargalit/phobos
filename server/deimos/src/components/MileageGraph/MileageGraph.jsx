import './MileageGraph.scss';

import { Spin } from 'antd';
import { meanBy as _meanBy } from 'lodash';
import moment from 'moment';
import React from 'react';
import {
  Area,
  AreaChart,
  CartesianGrid,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';

import AngledGraphTick from '../AngledGraphTick';

const transform = data =>
  data
    .sort((a, b) => moment(new Date(a.month)) - moment(new Date(b.month)))
    .map(({ month, miles }) => ({
      month,
      miles: parseFloat(miles.toFixed(2)),
    }));

const calculateAverage = data => _meanBy(data, 'miles');

export default function MileageGraph({ loading, monthlyData }) {
  if (loading) return <Spin />;
  const data = transform(monthlyData);

  return (
    <div className="activity-graph-wrapper">
      <div className="graph-header">
        <h2>M O N T H L Y</h2>
        <h2>M I L E A G E</h2>
      </div>
      <ResponsiveContainer width="100%" height={450}>
        <AreaChart
          className="activity-graph"
          data={data}
          margin={{ top: 10, right: 30, left: 0, bottom: 0 }}
          padding={{ top: 10, right: 30, left: 30, bottom: 10 }}
        >
          <defs>
            <linearGradient id="mileageColor" x1="0" y1="0" x2="0" y2="1">
              <stop offset="5%" stopColor="#d4504f" stopOpacity={0.6} />
              <stop offset="95%" stopColor="#d4504f" stopOpacity={0} />
            </linearGradient>
          </defs>
          <XAxis
            interval={2}
            dataKey="month"
            height={120}
            tick={<AngledGraphTick />}
          />
          <YAxis />
          <CartesianGrid strokeDasharray="3 3" />
          <ReferenceLine
            y={calculateAverage(data)}
            stroke="red"
            strokeDasharray="3 3"
            label={{
              position: 'top',
              fontWeight: 600,
              value: 'Monthly Average',
            }}
          />
          <Tooltip
            separator={null}
            formatter={value => [`${value} miles`, '']}
            animationDuration={300}
          />
          <Area
            dataKey="miles"
            type="monotone"
            stroke="#912827"
            fillOpacity={1}
            fill="url(#mileageColor)"
          />
        </AreaChart>
      </ResponsiveContainer>
    </div>
  );
}
