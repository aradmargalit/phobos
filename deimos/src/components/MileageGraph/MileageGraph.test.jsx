import { shallow } from 'enzyme';
import React from 'react';

import MileageGraph from './MileageGraph';

const generateComponent = props => {
  const defaultProps = {};

  const mergedProps = { ...defaultProps, ...props };

  return shallow(<MileageGraph {...mergedProps} />);
};

describe('<MileageGraph />', () => {
  it('matches snapshots', () => {
    expect(
      generateComponent({
        monthlyData: [{ month: 'January 2020', miles: 123 }],
      })
    ).toMatchSnapshot();
  });
});
