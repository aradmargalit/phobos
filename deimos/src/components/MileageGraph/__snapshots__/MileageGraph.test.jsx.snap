// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MileageGraph /> matches snapshots 1`] = `ShallowWrapper {}`;
